File: swift/MahjongScoreApp/Models/GameRecord.swift
import SwiftData
import Foundation

@Model
final class PlayerGameScore {
    var id: UUID = UUID()
    var game: GameRecord?
    
    @Relationship(deleteRule: .nullify)
    var player: Player?
    
    var rawScore: Int = 0
    var finalScore: Double = 0.0
    var rank: Int = 1
    var chipCount: Int = 0
    
    // Seat order is only needed if there was a tie to break. Optional.
    var seatOrder: Int? = nil 
    
    init(player: Player?, rawScore: Int, seatOrder: Int? = nil) {
        self.id = UUID()
        self.player = player
        self.rawScore = rawScore
        self.seatOrder = seatOrder
        self.chipCount = 0 // Chips moved to session, kept here 0 for backward logic compatibility if needed or removed
    }
}

@Model
final class GameRecord {
    var id: UUID = UUID()
    var timestamp: Date = Date()
    var dailySession: DailySession?
    
    @Relationship(deleteRule: .cascade, inverse: \PlayerGameScore.game)
    var playerScores: [PlayerGameScore] = []
    
    init(timestamp: Date = Date()) {
        self.id = UUID()
        self.timestamp = timestamp
    }
}

File: web-simulator/src/models/GameRecord.js
// GameRecord & PlayerGameScore - mirrors Swift models

export class PlayerGameScore {
  constructor(playerId, rawScore, seatOrder = null) {
    this.id = crypto.randomUUID();
    this.gameId = null;
    this.playerId = playerId;
    this.rawScore = rawScore;
    this.finalScore = 0.0;
    this.rank = 1;
    this.chipCount = 0;
    this.seatOrder = seatOrder;
  }
}

export class GameRecord {
  constructor(timestamp = new Date()) {
    this.id = crypto.randomUUID();
    this.timestamp = timestamp;
    this.dailySessionId = null;
    this.playerScores = []; // PlayerGameScore[]
  }
}
